pkg/resiliency/retry.go

```diff
@@ -0,0 +1,152 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resiliency
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/dapr/kit/retry"
+)
+
+type CodeError struct {
+	StatusCode int32
+	err        error
+}
+
+func (ce CodeError) Error() string {
+	return ce.err.Error()
+}
+
+func (ce CodeError) Unwrap() error {
+	return ce.err
+}
+
+func NewCodeError(statusCode int32, err error) CodeError {
+	return CodeError{
+		StatusCode: statusCode,
+		err:        err,
+	}
+}
+
+type Retry struct {
+	*retry.Config
+	*StatusCodeFilter
+}
+
+func NewRetry(retryConfig *retry.Config, statusCodeFilter *StatusCodeFilter) *Retry {
+	return &Retry{
+		Config:           retryConfig,
+		StatusCodeFilter: statusCodeFilter,
+	}
+}
+
+type StatusCodeFilter struct {
+	retryOnPatterns  []*regexp.Regexp
+	ignoreOnPatterns []*regexp.Regexp
+	defaultRetry     bool
+}
+
+func NewStatusCodeFilter() *StatusCodeFilter {
+	return &StatusCodeFilter{defaultRetry: true}
+}
+
+func ParseStatusCodeFilter(retryOnPatterns []string, ignoreOnPatterns []string) (*StatusCodeFilter, error) {
+	if len(retryOnPatterns) != 0 && len(ignoreOnPatterns) != 0 {
+		return nil, errors.New("retryOn and ignoreOn cannot be used together")
+	}
+	filter := NewStatusCodeFilter()
+	if err := filter.ParseRetryOnList(retryOnPatterns); err != nil {
+		return nil, err
+	}
+	if err := filter.ParseIgnoreOnList(ignoreOnPatterns); err != nil {
+		return nil, err
+	}
+	return filter, nil
+}
+
+func (f *StatusCodeFilter) ParseRetryOnList(patterns []string) error {
+	if len(patterns) == 0 {
+		return nil
+	}
+	regexpPatterns, err := f.parsePatterns(patterns)
+	if err != nil {
+		return err
+	}
+	f.retryOnPatterns = regexpPatterns
+	f.defaultRetry = false
+	return nil
+}
+
+func (f *StatusCodeFilter) parsePatterns(patterns []string) ([]*regexp.Regexp, error) {
+	regexpPatterns := make([]*regexp.Regexp, len(patterns))
+	for i, pattern := range patterns {
+		regexpPattern, err := compilePattern(pattern)
+		if err != nil {
+			return nil, err
+		}
+		regexpPatterns[i] = regexpPattern
+	}
+	return regexpPatterns, nil
+}
+
+func (f *StatusCodeFilter) ParseIgnoreOnList(patterns []string) error {
+	if len(patterns) == 0 {
+		return nil
+	}
+	regexpPatterns, err := f.parsePatterns(patterns)
+	if err != nil {
+		return err
+	}
+	f.ignoreOnPatterns = regexpPatterns
+	f.defaultRetry = true
+	return nil
+}
+
+func (f *StatusCodeFilter) StatusCodeNeedRetry(statusCode int32) bool {
+	statusCodeStr := strconv.Itoa(int(statusCode))
+
+	// Check whitelist patterns
+	for _, pattern := range f.retryOnPatterns {
+		if pattern.MatchString(statusCodeStr) {
```

Comment:
These are using regexps which are kind of slow. Could we use ranges rather than regexps? Since they are numbers?

Reply:
Yes, this makes sense. How about using the following patterns:
- a specific http code, such as "500"
- http code range, such as "500-509", 500 and 509 are included.
- *now can be removed, since 5** can be defined as "500-599"

Reply:
I think ranges are even clearer and more powerful. So, it could be:

- Precise status code: `500`
- Multiple status codes: `404,412` (comma-separated)
- Ranges: `400-499`
- A combination of the two above: `404,500-599`

Please let's do the parsing of the rule only once during initialization of the resiliency policies, and not on each request.